### ISSUE TITLE
py_image doesn't work with binaries that depend on other binaries

### ISSUE THREAD

**Issue by drigz** [bazelbuild/rules_docker#207]
If you have a binary than needs to start another binary, and you add the second binary to the first with `data`, it breaks `py_image` by causing it to add a non-existent "middleman runfile" to the files that `build_tar` expects.

```
  File ".../io_bazel_rules_docker/container/build_tar.py", line 124, in add_file
    gname=names[1])
  File ".../io_bazel_rules_docker/container/build_tar.runfiles/bazel_tools/tools/build_defs/pkg/archive.py", line 279, in add_file
    with open(file_content, 'rb') as f:
IOError: [Errno 2] No such file or directory: 'bazel-out/local-fastbuild/internal/_middlemen/foo-runfiles'
```

This appears to be linked to bazelbuild/bazel#2088, which was fixed in [a5294e6a](https://github.com/bazelbuild/bazel/commit/a5294e6a48a0c6ff6ef5eec918489f8564b40824) by replacing `getRunfiles().getArtifacts()` with `getRunfilesArtifactsWithoutMiddlemen()`.

I have [a workaround](https://github.com/drigz/rules_docker/commit/39ced16812f288120fa495d7cc033f2b4bc8f102) - feel free to copy it or ask me to send a PR - but hopefully there's a less hacky way to fix it.

### Repro

```
> cat WORKSPACE 
git_repository(
    name = "io_bazel_rules_docker",
    commit = "58d022892232e5d59daba7760289976d5f6e7433",
    remote = "https://github.com/bazelbuild/rules_docker.git",
)

load(
    "@io_bazel_rules_docker//python:image.bzl",
    _py_image_repos = "repositories",
)

_py_image_repos()
> cat BUILD.bazel 
load("@io_bazel_rules_docker//python:image.bzl", "py_image")

py_binary(
    name = "foo",
    srcs = ["foo.py"],
)

py_image(
    name = "main",
    srcs = ["main.py"],
    data = [":foo"],
    main = "main.py",
)
> echo > foo.py
> echo > main.py
> bazel build :main
INFO: Analysed target //:main (0 packages loaded).
INFO: Found 1 target...
ERROR: /usr/local/google/home/rodrigoq/git/bazeltest/data_py_binary/BUILD.bazel:8:1: ImageLayer main-layer.tar failed (Exit 1)
Traceback (most recent call last):
  File "/usr/local/google/home/rodrigoq/.cache/bazel/_bazel_rodrigoq/e8a75b06501f3bcce14a5ebd69ff6fce/bazel-sandbox/1852124494563052974/execroot/__main__/bazel-out/host/bin/external/io_bazel_rules_docker/container/build_tar.runfiles/__main__/../io_bazel_rules_docker/container/build_tar.py", line 274, in <module>
    main(FLAGS(sys.argv))
  File "/usr/local/google/home/rodrigoq/.cache/bazel/_bazel_rodrigoq/e8a75b06501f3bcce14a5ebd69ff6fce/bazel-sandbox/1852124494563052974/execroot/__main__/bazel-out/host/bin/external/io_bazel_rules_docker/container/build_tar.runfiles/__main__/../io_bazel_rules_docker/container/build_tar.py", line 261, in main
    output.add_file(inf, tof, mode, ids, names)
  File "/usr/local/google/home/rodrigoq/.cache/bazel/_bazel_rodrigoq/e8a75b06501f3bcce14a5ebd69ff6fce/bazel-sandbox/1852124494563052974/execroot/__main__/bazel-out/host/bin/external/io_bazel_rules_docker/container/build_tar.runfiles/__main__/../io_bazel_rules_docker/container/build_tar.py", line 124, in add_file
    gname=names[1])
  File "/usr/local/google/home/rodrigoq/.cache/bazel/_bazel_rodrigoq/e8a75b06501f3bcce14a5ebd69ff6fce/bazel-sandbox/1852124494563052974/execroot/__main__/bazel-out/host/bin/external/io_bazel_rules_docker/container/build_tar.runfiles/bazel_tools/tools/build_defs/pkg/archive.py", line 279, in add_file
    with open(file_content, 'rb') as f:
IOError: [Errno 2] No such file or directory: 'bazel-out/local-fastbuild/internal/_middlemen/foo-runfiles'
Target //:main failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.531s, Critical Path: 0.19s
FAILED: Build did NOT complete successfully
```

**Comment by mattmoor** [comment]
I want to make sure that I understand.  There is a Bazel bug (fixed at HEAD), and the PR would be a workaround?

**Comment by drigz** [comment]
No, there was a Bazel bug in `build_python_zip`, where it tried to include non-existent files in the archive, which was fixed on 0.4.5. There is an analogous bug that shows up `_app_layer_impl`. It's not clear to me if `_app_layer_impl` is at fault, or if Bazel's implementation of [default_runfiles](https://docs.bazel.build/versions/master/skylark/lib/globals.html#DefaultInfo) is at fault.

I've filed bazelbuild/bazel#3994 upstream to check.

**Comment by drigz** [comment]
Looks like this will be fixed in the next Bazel release.

**Comment by drigz** [comment]
Fixed in bazel 0.8.0.
